[ENH] Add nan policy handler for FallbackForecaster #5897

ninedigits opened the pull request:
<!--
Welcome to sktime, and thanks for contributing!
Please have a look at our contribution guide:
https://www.sktime.net/en/latest/get_involved/contributing.html
-->

#### Reference Issues/PRs
-Issue: https://github.com/sktime/sktime/issues/5897
-Additional discussion: https://github.com/sktime/sktime/pull/5847


#### What does this implement/fix? Explain your changes.

We've introduced a `nan_predict_policy` within the `FallbackForecaster` class to handle NaN values in forecaster predictions. This policy defaults to "ignore" and can be set to "raise", which will actively check for NaN values in the prediction output and raise an exception if any are detected, or "warn", which will warn the user if any values are detected. This enhancement is aimed at improving the robustness and reliability of the forecasting process, especially when integrating transformations like BoxCox with ARIMA forecasters.


#### Does your contribution introduce a new dependency? If yes, which one?

No new dependencies.

#### What should a reviewer concentrate their feedback on?

- Correctness of the NaN detection and exception raising mechanism.
- Adhering to existing `sktime` design standards.
- Adequacy of tests covering the new functionality.

#### Did you add any tests for the change?

Yes, tests have been added to ensure the `nan_predict_policy` behaves as expected in scenarios where NaN values are predicted. These tests cover both the "raise" and "warn" policies to verify their correct implementation.

#### Any other comments?
<!--
We value all user contributions, no matter how small or complex they are. If you have any questions, feel free to post
in the dev-chat channel on the sktime discord https://discord.com/invite/54ACzaFsn7. If we are slow to review (>3 working days), likewise feel free to ping us on discord. Thank you for your understanding during the review process.
-->

#### PR checklist
<!--
Please go through the checklist below. Please feel free to remove points if they are not applicable.
-->

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [ ] Optionally, I've added myself and possibly others to the [CODEOWNERS](https://github.com/sktime/sktime/blob/main/CODEOWNERS) file - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [ ] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [ ] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [ ] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [ ] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency).

<!--
Thanks for contributing!
-->

fkiraly commented on sktime/forecasting/compose/_fallback.py:
```diff
@@ -36,6 +47,13 @@ class is useful in scenarios where the reliability of individual forecasting mod
 
+    nan_predict_policy: str, default='ignore'
+        Determines the action to take if NaN values are found in the predictions. If set
+         to 'raise', the method raises a ValueError when NaN values are detected in the
```
is this indentation change intented? Might render strangely.

ninedigits replied:
Good catch

fkiraly replied:
I mean, is the intentation indented, no, the intendation intented - well, you know what I mean

ninedigits replied:
I understood what you meant, the indentation was wrong. Was a good catch!

fkiraly replied:
I was just joking about the typo I made

fkiraly commented on sktime/forecasting/compose/_fallback.py:
```diff
@@ -36,6 +47,13 @@ class is useful in scenarios where the reliability of individual forecasting mod
 
+    nan_predict_policy: str, default='ignore'
+        Determines the action to take if NaN values are found in the predictions. If set
+         to 'raise', the method raises a ValueError when NaN values are detected in the
```
I would explain `raise` better - the exception is in fact not raised to hte user, but handled by this class, i.e., occurrence of nans is considered a failure for the fallback

fkiraly commented on sktime/forecasting/compose/_fallback.py:
```diff
@@ -96,14 +114,17 @@ class is useful in scenarios where the reliability of individual forecasting mod
         self.verbose = verbose
 
+        _check_nan_policy_option(nan_predict_policy)
+        self.nan_predict_policy = nan_predict_policy
```
slightly more idiomatic (similar ot sklearn): have the checker return the input, then do `self.palim = check_palim(palim)`

fkiraly commented on sktime/forecasting/compose/_fallback.py:
```diff
@@ -112,6 +133,20 @@ def __init__(self, forecasters, verbose=False):
 
+    def _validate_y_pred(self, y_pred):
+        if self.nan_predict_policy in ("warn", "raise"):
+            num_nans = y_pred.isna().sum()
```
* should this use `isnull`? Or should it be an option to use `isna` vs `isnull`?
* `any` is faster than `sum` in this case

fkiraly reviewed (changes requested):
Seems to work, and it is tested.

I've added some comments, but would consider none of these blocking.

ninedigits commented on sktime/forecasting/compose/_fallback.py:
```diff
@@ -112,6 +133,20 @@ def __init__(self, forecasters, verbose=False):
+            num_nans = y_pred.isna().sum()
+            if num_nans > 0:
+                msg = f"Null value presents in predict: {y_pred}"
+                if self.nan_predict_policy == "raise":
```
@fkiraly Isn't this handled in line 137?
```python3
    def _validate_y_pred(self, y_pred):
        if self.nan_predict_policy in ("warn", "raise"):
            num_nans = y_pred.isna().sum() ### todo, change this to isnull().any()
```

fkiraly replied:
oh, yes, it is.

ninedigits commented:
> Seems to work, and it is tested.
> 
> I've added some comments, but would consider none of these blocking.

Good revisions suggestions, I added them in.

ninedigits commented:
In the future, we can add to `_validate_y_pred()` to handle other non-exception forecasting errors. E.g. exploding forecasts, although that one could be tricky to implement.

fkiraly commented on sktime/forecasting/compose/_fallback.py:
```diff
@@ -36,6 +48,22 @@ class is useful in scenarios where the reliability of individual forecasting mod
+    nan_predict_policy: str, default='ignore'
+        Determines the action to take if NaN values are found in the predictions.
+        Available options:
+        * "ignore"
```
I think this does not render right if there is no space before and after the bullet point list

fkiraly reviewed (approved):
Thanks - one non-blocking comment above about a formatting problem in the docstring.

fkiraly commented:
> In the future, we can add to `_validate_y_pred()` to handle other non-exception forecasting errors. E.g. exploding forecasts, although that one could be tricky to implement.

Very interesting idea. Perhaps it is worthwhile to think about how to make this abstract? E.g., allow any parameter estimator to define the condition? Another place where sth like this exists is the `TransformIf` transformer, so I wonder whether we can split off a "condition" object or parameter estimator?

ninedigits commented:
> Thanks - one non-blocking comment above about a formatting problem in the docstring.

Committed those changes

ninedigits commented:
> > In the future, we can add to `_validate_y_pred()` to handle other non-exception forecasting errors. E.g. exploding forecasts, although that one could be tricky to implement.
> 
> Very interesting idea. Perhaps it is worthwhile to think about how to make this abstract? E.g., allow any parameter estimator to define the condition? Another place where sth like this exists is the `TransformIf` transformer, so I wonder whether we can split off a "condition" object or parameter estimator?

Like a parameter where a function or estimator is passed in capable of assessing the inputs and outputs, and allowing the user to specify their own validation params?

fkiraly commented:
> Like a parameter where a function or estimator is passed in capable of assessing the inputs and outputs, and allowing the user to specify their own validation params?

Yes! The closest type of such estimator/object is the "parameter estimator", currently used mostly for stationarity and seasonality detection, the scitype code is `param_est` (for use in `all_estimators`). New issue to discuss?

ninedigits commented:
> Yes! The closest type of such estimator/object is the "parameter estimator", currently used mostly for stationarity and seasonality detection, the scitype code is `param_est` (for use in `all_estimators`). New issue to discuss?

It's a good idea, I can take a look at it. I'm wondering if it would be it's own estimator that would wrap around an input forecaster within FallbackForecaster?

fkiraly commented:
good question - I think there's a deeper API design discussion her.e

Opened an issue to move it out of this PR which will get merged soon: https://github.com/sktime/sktime/issues/5936

Simple reason, merged PR are very low visibility, whereas open issues signify ongoing work.
